src/font/sdl_ttf.cpp

```diff
@@ -43,122 +43,82 @@ static lg::log_domain log_font("font");
 #define WRN_FT LOG_STREAM(warn, log_font)
 #define ERR_FT LOG_STREAM(err, log_font)
 
-namespace font {
-
-/***
- * Caches used to speed up font rendering
- */
-
+namespace font
+{
+namespace
+{
 // Record stored in the font table.
 // If the record for font_id (FOO, Bold + Underline) is a record (BAR, Bold),
 // it means that BAR is a Bold-styled version of FOO which we shipped with the
 // game, and now SDL_TTF should be used to style BAR as underline for the final results.
 struct ttf_record
 {
-	TTF_Font* font;
+	std::shared_ptr<TTF_Font> font;
 	int style;
 };
-
 static std::map<font_id, ttf_record> font_table;
+
+// The indices in these vectors correspond to the font_id.subset values in font_table.
 static std::vector<std::string> font_names;
 static std::vector<std::string> bold_names;
 static std::vector<std::string> italic_names;
 
-struct char_block_map
+struct family_record
 {
-	char_block_map()
-		: cbmap()
-	{
-	}
+	std::shared_ptr<const TTF_Font> font;
+	subset_id subset;
+	std::string name;
+};
+/**
+ * Used for implementing find_font_containing, the elements are in the same order as the arguments
+ * to set_font_list(). The fonts here are a subset of those in font_table, because
+ * find_font_containing doesn't need size-specific instances of a font.
+ *
+ * In most locales, the subset_ids will match the indices into this vector.  This is only a
+ * coincidence, and it won't be true (at the time of writing) in Chinese.
+ *
+ * \todo Are all variants of a font guaranteed to have exactly the same glyphs? For example, might
+ * an italic variant only contain the glyphs which are major improvements on an automatic skew of
+ * the non-italic version?
+ */
+std::vector<family_record> family_table;
 
-	typedef std::pair<int, subset_id> block_t;
-	typedef std::map<int, block_t> cbmap_t;
-	cbmap_t cbmap;
-	/** Associates not-associated parts of a range with a new font. */
-	void insert(int first, int last, subset_id id)
-	{
-		if (first > last) return;
-		cbmap_t::iterator i = cbmap.lower_bound(first);
-		// At this point, either first <= i->first or i is past the end.
-		if (i != cbmap.begin()) {
-			cbmap_t::iterator j = i;
-			--j;
-			if (first <= j->second.first /* prev.last */) {
-				insert(j->second.first + 1, last, id);
-				return;
-			}
-		}
-		if (i != cbmap.end()) {
-			if (/* next.first */ i->first <= last) {
-				insert(first, i->first - 1, id);
-				return;
-			}
-		}
-		cbmap.emplace(first, block_t(last, id));
-	}
-	/**
-	 * Compresses map by merging consecutive ranges with the same font, even
-	 * if there is some unassociated ranges in-between.
```

Comment:
So, this bit of functionality here was the cause of the bug?

Reply:
Yup. I think I found more bugs as the refactor grew, but this was the base of the Chinese characters being invisible.